xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -301,12 +301,22 @@ void watchClusterData(String clusterName, ClusterWatcher watcher) {
     // If local cache contains cluster information to be watched, notify the watcher immediately.
     if (absentCdsResources.contains(clusterName)) {
       logger.log(XdsLogLevel.DEBUG, "Cluster resource {0} is known to be absent", clusterName);
-      watcher.onResourceDoesNotExist(clusterName);
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          watcher.onResourceDoesNotExist(clusterName);
+        }
+      });
       return;
     }
     if (clusterNamesToClusterUpdates.containsKey(clusterName)) {
       logger.log(XdsLogLevel.DEBUG, "Retrieve cluster info {0} from local cache", clusterName);
-      watcher.onClusterChanged(clusterNamesToClusterUpdates.get(clusterName));
+      syncContext.execute(new Runnable() {
+        @Override
+        public void run() {
+          watcher.onClusterChanged(clusterNamesToClusterUpdates.get(clusterName));
```

Comment:
Do you need to check clusterNamesToClusterUpdates.containsKey(clusterName) again?

Reply:
Or just make a final copy of `clusterNamesToClusterUpdates.get(clusterName)` before `syncContext.execute()`